react/features/mobile/callkit/middleware.js

```diff
@@ -0,0 +1,194 @@
+            }
+        }
+
+        break;
+    }
+
+    case APP_WILL_MOUNT: {
+        CallKit.setup();  // TODO: set app icon.
+        const listeners = new Map();
+        const callEndListener = data => {
+            const conference = getCurrentConference(getState);
+
+            if (conference && conference.callUUID === data.callUUID) {
+                // We arrive here when a call is ended by the system, for
+                // for example when another incoming call is received and the
+                // user selects "End & Accept".
+                delete conference.callUUID;
+                dispatch(appNavigate(undefined));
+            }
+        };
+
+        listeners.set('performEndCallAction', callEndListener);
+
+        // Set the same listener for providerDidReset. According to the docs,
+        // when the system resets we should terminate all calls.
+        listeners.set('providerDidReset', callEndListener);
+
+        const setMutedListener = data => {
+            const conference = getCurrentConference(getState);
+
+            if (conference && conference.callUUID === data.callUUID) {
+                // Break the loop. Audio can be muted both from the CallKit
+                // interface and from the Jitsi Meet interface. We must keep
+                // them in sync, but at some point the loop needs to be broken.
+                // We are doing it here, on the CallKit handler.
+                const { muted } = getState()['features/base/media'].audio;
+
+                if (muted !== data.muted) {
+                    dispatch(setAudioMuted(Boolean(data.muted)));
+                }
+
+            }
+        };
+
+        listeners.set('performSetMutedCallAction', setMutedListener);
+
+        dispatch({
+            type: _SET_CALLKIT_LISTENERS,
+            listeners
+        });
+        break;
+    }
+
+    case APP_WILL_UNMOUNT:
+        dispatch({
+            type: _SET_CALLKIT_LISTENERS,
+            listeners: null
+        });
+        break;
+
+    case CONFERENCE_FAILED: {
+        const { callUUID } = action.conference;
+
+        if (callUUID) {
+            CallKit.reportCallFailed(callUUID);
+        }
+
+        break;
+    }
+
+    case CONFERENCE_LEFT: {
+        const { callUUID } = action.conference;
+
+        if (callUUID) {
+            CallKit.endCall(callUUID);
+        }
+
+        break;
+    }
+
+    case CONFERENCE_JOINED: {
+        const { callUUID } = action.conference;
+
+        if (callUUID) {
+            CallKit.reportConnectedOutgoingCall(callUUID);
+        }
+
+        break;
+    }
+
+    case CONFERENCE_WILL_JOIN: {
+        const conference = action.conference;
+        const url = getInviteURL(getState);
+        const hasVideo = !isVideoMutedByAudioOnly({ getState });
+
+        // When assigning the call UUID, do so in upper case, since iOS will
+        // return it upper cased.
+        conference.callUUID = uuid.v4().toUpperCase();
```

Comment:
I went back-and-forth on the approach here, and settled on storing the UUID attached to the Conference object itself instead of the state. The reason being, we may have a joining and an ending conference, and it's easier to keep track of everything if each conference object has a UUID itself.